docs/dev/system_keyspace.md

```diff
@@ -149,6 +149,51 @@ merge the data in the legacy store with data the `truncated` table. Until the wh
 cluster agrees on the feature `TRUNCATION_TABLE` truncation will write both new and 
 legacy records. When the feature is agreed upon the legacy map is removed.
 
+## system.quarantined_hosts
+
+Holds a list of host identifiers of nodes that were removed from the cluster and are not allowed to join back.
+
+Node operations like remove node, replace node, or decommissioned reassign token ownership of the target node.
+After such operations, the removed host must not rejoin the cluster as it might hold stale data, causing data resurrection.
+
+Schema:
+~~~
+CREATE TABLE system.quarantined_hosts (
+    key text,
+    host_id uuid,
+    quarantined_at timestamp,
+    PRIMARY KEY (key, host_id)
+) WITH CLUSTERING ORDER BY (host_id ASC)
+~~~
+
+The table consists of a single partition (where `key = 'QUARANTINED_HOSTS'`)
+and each quarantined host is stored in a row, with the `host_id` as the clustering key.
+The time when the host was quarantined is stored in `quarantined_at`.
+
+When node operations like `removenode`, `replace`, or `decommission` complete,
+the target hosts are quarantined from the token ring.
+Their `host_id` is inserted to `system.quarantined_hosts` and the `QUARANTINED_HOSTS` application state is updated and gossiped to all other nodes
```

Comment:
Storing all the quarantined hosts into QUARANTINED_HOSTS is not very nice. QUARANTINED_HOSTS could grow large O(n). However, this thing does not very often though (unlike the per table hit rate).

Maybe it is better to use a distributed table for this.

Reply:
> Storing all the quarantined hosts into QUARANTINED_HOSTS is not very nice. QUARANTINED_HOSTS could grow large O(n).

This contradicts what you said above:
> > Even if we have 10000 nodes get removed / replaced in a cluster, this data is tiny.

> However, this thing does not very often though (unlike the per table hit rate).
> 
> Maybe it is better to use a distributed table for this.

I thought so too and tried to implement that, but we have a chicken and egg problem, and system distributed is started only later on and we need to access it earlier in the boot process.

Reply:
A raft table would be ideal, but with raft, we probably wouldn't need system.quarantined_hosts at all...

Reply:
It does not contradict. It is small for what we can store in a table but it does not mean it is small to store in a single rpc message.

Reply:
Are you worried about the gossip messages or the node ops message?
In the latter case it would be a single host at most at the moment.
In the former case, yes, we'd send all known quarantined hosts, hence the TTL, so not to accumulate quarantined hosts forever.

Reply:
I worried about gossip message goes big. However, this is not a reason to ttl the quarantined host list. If we have a list, It is important to keep it forever to avoid having admin must do something in 30 days.

Reply:
> I worried about gossip message goes big. However, this is not a reason to ttl the quarantined host list. If we have a list, It is important to keep it forever to avoid having admin must do something in 30 days.

It's like saying we need to keep tombstones forever and never purge them since the admin can make a mistake and cause data resurrection.

> Guys, the patches are good enough. Please don't try to make them perfect - - it's an interim solution for a couple of releases before people will default on raft. @asias this is about your comments on quarantined_hosts table and rpc and handle_state_normal.

Agreed.  If there are real issue with the series let's address them.
If there are more ideas for improvements, I suggest to merge first as this already provides a nice improvement, and continue the discussion towards a follow-up, ideally achieving the goals using raft.

Reply:
We understand the solution doesn't cover all cases. That's fine. As @kostja said, it's interim.
We need this to solve our current problems at hand. It's really blocking some of our plans. We need to get this merged ASAP.

Reply:
What we need for raft is only 1), no?

Reply:
True. But if I understand correctly, 2) prevents a regression (modulo the 30 days TTL...) introduced by 1) - see https://github.com/scylladb/scylladb/pull/12084#issuecomment-1341219742.

Reply:
> True. But if I understand correctly, 2) prevents a regression (modulo the 30 days TTL...) introduced by 1) - see [#12084 (comment)](https://github.com/scylladb/scylladb/pull/12084#issuecomment-1341219742).

Right. https://github.com/scylladb/scylladb/blob/a076ceef97d505d88e80b8504750f024c9a976a2/service/storage_service.cc#L925

doesn't work any more.

Reply:
@bhalevy but I don't think this collision mechanism was in any way reliable - couldn't the old (replaced) node restart, causing the gossip generation number to increase, causing it to win the collision?

I.e., I'm trying to see that maybe there was no real regression, because the mechanism was broken all this time anyway.

Reply:
Even so, I think that the interim quarantine solution is worth while.

Reply:
> Maybe. The naive dtest did restart the replaced node and declared success once the collision message was found on the other nodes, but it didn't let the scenario complete and make sure that the restarted replaced node wasn't able to reclaim its tokens back.

I ran the test. It doesn't make sense.
This code:
```python3
        # check that restarting node 3 doesn't work
        # FIXME: https://github.com/scylladb/scylla/issues/5523 is fixed
        # need to verify that the node doesn't start listening
        logger.info("Try to restart node 3 (should fail)")
        node3.start(no_wait=True)
        checkCollision = node1.grep_log("between .*"+self.cluster.get_node_ip(3)+" and .*" +
        logger.info(checkCollision)
        assert len(checkCollision) == 1
```

First of all, the `checkCollision` assertion passes, but it has nothing to do with node3's restart. The log is printed earlier, when **node 4 bootstraps and replaces node 3**.
Node 3 restart doesn't even complete. The test does not wait for that `no_wait=True`. The startup gets interrupted because the test is finishing.

If we actually wait for node 3 to restart (`no_wait=False`), it steals its host ID back...
The test hangs for some time - I don't know why, it's as if node 3 is not starting (but it has already started, I checked the logs, maybe it's not making CQL available for some reason or sth), but eventually it fails because node 4 does `on_internal_error`:
```
ERROR 2022-12-08 14:16:25,899 [shard 0] storage_service - endpoint=127.0.94.3 is marked for removal but still owns 256 tokens, at: 0x35c50ee 0x35c56a0 0x35c5988 0x3467473 0x34675f8 0x2123ba0 0x1022e9d 0x348f4f4 0x34905e7 0x348fa99 0x3443d23 0x3442f98 0xf9a9ce 0xfc9d50 0xf99ca7 /lib64/libc.so.6+0x2750f /lib64/libc.so.6+0x275c8 0xf98fe4
```
(I think this is a consequence of c61083852cb68c504a9d487ba0d7885af19cce1b)

Reply:
Anyway, one conclusion we can make is: the stealing of Host ID by the replacing node did not prevent the replaced node from coming back. So **removing the stealing does not introduce a regression**. The test is failing because the `checkCollision` assertion is failing. And that obviously happens because the test is actually checking whether the replacing node steals the replaced node's Host ID (which we're aware of and consciously decided to change).

Reply:
> (I think this is a consequence of https://github.com/scylladb/scylladb/commit/c61083852cb68c504a9d487ba0d7885af19cce1b)

Happens without that commit as well.

So if we wait for 3 node to restart, it will cause node 4 to crash. And this happens without any of your changes.

Reply:
> The test hangs for some time - I don't know why

That's because
1. `no_wait=False` implies `wait_other_notice=True`, which means that `node.start(...)` will return only when all other nodes print `InetAddress <addr> is now UP`
2. node 3 was just replaced with node 4, causing other nodes to place a "gossip quarantine" on node 3 for 60 seconds (that's an existing mechanism)
3. the test gets unstuck once this gossip quarantine elapses and the nodes notice that node 3 is UP

```
INFO  2022-12-08 14:39:09,624 [shard 0] gossip - 60000 ms elapsed, 127.0.3.3 gossip quarantine over
INFO  2022-12-08 14:39:10,625 [shard 0] gossip - InetAddress 127.0.3.3 is now UP, status = NORMAL
```